text/3806-align-attr.md

```diff
@@ -0,0 +1,539 @@
+    pub x: u8,
+    #[align(128)]
+    pub y: u8,
+    pub z: u8,
+}
+```
+
+```rust
+#[repr(linear, align(128))]
+pub struct foo2 {
+    pub x: u8,
+    pub _padding: [MaybeUninit<u8>; 127usize],
+    pub y: u8,
+    pub z: u8,
+}
+```
+
+## On `static`s
+
+Any `static` item (including `static`s inside `extern` blocks) may have an
+`align` attribute applied:
+
+```rust
+#[align(32)]
+static BAZ: [u32; 12] = [0xDEADBEEF; 12];
+
+unsafe extern "C" {
+    #[align(2)]
+    safe static BOZZLE: u8;
+}
+```
+
+The effect of the attribute is to force the `static` to be stored with at least
+the specified alignment. The attribute does not force padding bytes to be added
+after the `static`. For `static`s inside `unsafe extern` blocks, if the `static`
+does not meet the specified alignment, the behavior is undefined. (This UB is
+analogous to the UB that can result if the static item is not a valid value of
+its type. The question of whether the UB can occur even if the item is unused,
+has the same answer for both cases.)
+
+The `align` attribute may also be applied to thread-local `static`s created with
+the `thread_local!` macro; the attribute affects the alignment of the underlying
+value, not that of the outer `std::thread::LocalKey`.
+
+```rust
+thread_local! {
+    #[align(64)]
+    static FOO: u8 = 42;
+}
+
+fn main() {
+    FOO.with(|r| {
+        let p: *const u8 = r;
+        assert_eq!(p.align_offset(64), 0);
+    });
+}
+```
+
+`align` attributes on `static`s are shown in `rustdoc`-generated documentation.
+
+## On function items
+
+On function items, `#[align(…)]` sets the alignment of the function’s code. (It
+does not affect the alignment of its function item type, which remains a 1-ZST.)
+This replaces `#[repr(align(…))]` on function items, from
+`#![feature(fn_align)]`.
+
+On `async fn`, the attribute controls the alignment of the code of the function
+that returns the `Future`.
+
+The numerical value of a function pointer to a function with an `#[align(n)]`
+attribute is *not* always guaranteed to be a multiple of `n` on all targets. For
+example, on 32-bit ARM, the low bit of the function pointer is set for functions
+using the Thumb instruction set, even though the actual code of the function is
+always aligned to at least 2 bytes.
+
+`align` attributes on function items are shown in `rustdoc`-generated
+documentation.
+
+## On local variables
+
+The `align` attribute may also be applied to local variable declarations inside
+`let` bindings. The attribute forces the local to have at least the alignment
+specified:
+
+```rust
+fn main() {
+    let (a, #[align(4)] b, #[align(2)] mut c) = (4u8, 2u8, 1u8);
+    c *= 2;
+    dbg!(a, b, c);
+
+    if let Some(#[align(4)] x @ 1..) = Some(42u8) {
+        dbg!(x);
+        let p: *const u8 = x;
+        assert_eq!(p.align_offset(4), 0);
+    }
+}
+```
+
+The `align` attribute may not be applied to function parameters.
+
+```rust
+fn foo(#[align(8)] _a: u32) {} //~ ERROR
+```
```

Comment:
Do you propose we disallow this syntactically or semantically?

Reply:
Good question! Semantically, RFC updated